### PR TITLE
Feature/pch overhaul

### LINKED ISSUE #1194: Precompiled headers with Qt6 break CMake
Tested with CMake 3.21.2 and Qt 6.1.3.

```cmake
CMake Error at cmake/cotire.cmake:2376 (file):
  Error evaluating generator expression:

    $<TARGET_PROPERTY:QT_NO_UNICODE_DEFINES>

  $<TARGET_PROPERTY:prop> may only be used with binary targets.  It may not
  be used with add_custom_command or add_custom_target.  Specify the target
  to read a property from using the $<TARGET_PROPERTY:tgt,prop> signature
  instead.
Call Stack (most recent call first):
  cmake/cotire.cmake:2995 (cotire_generate_target_script)
  cmake/cotire.cmake:3385 (cotire_process_target_language)
  cmake/cotire.cmake:3569 (cotire_target)
  cmake/precompileheaders.cmake:142 (cotire)
  modules/qtwidgets/CMakeLists.txt:209 (ivw_compile_optimize_on_target)
```

see offending target property: https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/Qt6CoreMacros.cmake#n2424

### PR DESCRIPTION
**Pull request by martinfalk**
Fixes #1194

Changes proposed in this PR:
 * switch PCH generation from Cotire to CMake (potential downside: pch file gets recreated every time CMake runs)
 * only use a single PCH file containing a number of "heavy" includes from Core

This has been tested on: MSVC 17.3.4, CMake 3.24.0

### CODE REVIEW

File: cmake/globalmacros.cmake [hunk 1]
```diff
@@ -562,3 +562,22 @@ function(ivw_add_external_projects)
         add_subdirectory(${project_root_path} ${CMAKE_CURRENT_BINARY_DIR}/ext_${FOLDER_NAME})
     endforeach()
 endfunction()
+
+#--------------------------------------------------------------------
+# Get the list of targets the given target depends on. 
+function(ivw_get_target_depending_on retval target alsoInterfaceTargets)
```
Comment (ResearchDaniel) [inline cmake/globalmacros.cmake]:
```suggestion
function(ivw_get_target_dependencies retval target alsoInterfaceTargets)
```

File: cmake/precompileheaders.cmake [hunk 2]
```diff
@@ -27,118 +27,55 @@
...
     ivw_compile_optimize_on_target(${target})
 endfunction()
 
-# Optimize compilation with pre-compilied headers
-# Custom target properties:
-#  * COTIRE_PREFIX_HEADER_PUBLIC_IGNORE_PATH  
-#  * COTIRE_PREFIX_HEADER_PUBLIC_INCLUDE_PATH 
-# We make sure that these properties are propagated to the 
-# depending targets.
+# Optimize compilation with reusing pre-compilied headers
```
Comment (ResearchDaniel) [inline cmake/precompileheaders.cmake]:
```suggestion
# Optimize compilation by reusing pre-compilied headers
```

File: cmake/precompileheaders.cmake [hunk 3]
```diff
@@ -27,118 +27,55 @@
 # 
 #################################################################################
 
-
-
-if(WIN32)
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
     option(IVW_CFG_PRECOMPILED_HEADERS "Create and use precompilied headers" ON)
```
Comment (ResearchDaniel) [inline cmake/precompileheaders.cmake]:
```suggestion
    option(IVW_CFG_PRECOMPILED_HEADERS "Compile faster using pre-compiled headers" ON)
```

File: src/core/CMakeLists.txt [hunk 4]
```diff
@@ -863,8 +863,37 @@ target_link_libraries(inviwo-core
         utf8::cpp
 )
 
-#Optimize compiliation with PCH etc
-ivw_compile_optimize_on_target(inviwo-core)
+set(PCH_HEADERS
```
Comment (ResearchDaniel) [inline src/core/CMakeLists.txt]: Would be nice with a comment on how to reason when adding/removing headers here

**Review by ResearchDaniel — DISMISSED** [review]
Looks good to me! Just some minor suggestions

File: cmake/globalmacros.cmake (hunk 1 above)
Comment (petersteneteg) [inline cmake/globalmacros.cmake]: Why not use ivw_get_target_property_recursive
It this function is needed it should be in the globalutils.cmake file...

File: cmake/precompileheaders.cmake [hunk 5]
```diff
@@ -27,118 +27,55 @@
...
+            target_precompile_headers(${target} ${ARG_VISIBILITY} ${ARG_HEADERS})
+            file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/pch-dummy.cpp 
+                "// Dummy source file needed in target for reusing pch artifacts\n"
+            )
+            add_library(${target}-pch STATIC ${CMAKE_CURRENT_BINARY_DIR}/pch-dummy.cpp)
+            ivw_define_standard_definitions(${target}-pch ${target}-pch)
+            ivw_define_standard_properties(${target}-pch)
+            target_link_libraries(${target}-pch PRIVATE ${target})
+            target_precompile_headers(${target}-pch ${ARG_VISIBILITY} ${ARG_HEADERS})
+            ivw_folder(${target}-pch "pch-targets")
```
Comment (petersteneteg) [inline cmake/precompileheaders.cmake]: What is all this about?

File: src/core/CMakeLists.txt (hunk 4 above)
Comment (petersteneteg) [reply src/core/CMakeLists.txt]: Yes, what about adding vector? unordered_(map/set) 
Why are you adding c headers?

**Review by petersteneteg — CHANGES_REQUESTED** [review]

File: src/core/CMakeLists.txt (hunk 4 above)
Comment (martinfalk) [reply src/core/CMakeLists.txt]: These are taken from the top list of includes taking the most time as measured with the CompileScore extension.

Comment (petersteneteg) [reply src/core/CMakeLists.txt]: what happens if you add more?

File: cmake/precompileheaders.cmake (hunk 5 above)
Comment (martinfalk) [reply cmake/precompileheaders.cmake]: See https://discourse.cmake.org/t/help-with-target-precompile-headers/1065 why a dummy target is needed.

File: src/core/CMakeLists.txt (hunk 4 above)
Comment (martinfalk) [reply src/core/CMakeLists.txt]: There seems to be no impact. Tried with `data.h`, `volume.h`, and `image.h` with no effect even though they accounted for quite some include times. Needs testing with additional standard headers.

File: cmake/precompileheaders.cmake (hunk 5 above)
Comment (petersteneteg) [reply cmake/precompileheaders.cmake]: that is some shady stuff, and seems mostly concerned with static linking and object libraries, neither which I think you are using here. So please elaborate on why this is needed?

**Comment by martinfalk** [comment]
Reworked the PCH generation.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,18 @@
 Here we document changes that affect the public API or changes that needs to be communicated to other developers. 
 
+## 2022-09-29 Pre-compiled headers overhaul
+The generation of pre-compiled headers (PCH) was changed from Cotire to CMake. This enables using PCH in combination with Qt6.
+A single PCH file is generated for inviwo core including a number of standard includes. In addition, there is a PCH file for the qtwidgets modules with some Qt headers. These PCH files ar reused by all modules by default when depending on qtwidgets or core respectively, and assuming `IVW_CFG_PRECOMPILED_HEADERS` is checked. Use the `NO_PCH` option when calling `ivw_create_module()` to prevent the default PCH from being used in a module.
```
Comment (petersteneteg) [inline CHANGELOG.md:5]: Not sure if this need updating, might explain the difference between using target_precompile_headers directly and ivw_compile_optimize_on_target

**Review by petersteneteg — DISMISSED** [review]

File: CHANGELOG.md (hunk 6 above)
Comment (martinfalk) [reply CHANGELOG.md:5]: good suggestion

**Review by petersteneteg — APPROVED** [review]

**Pull request merged**